synthtool/gcp/templates/metadata.py

```diff
@@ -0,0 +1,185 @@
+
+    def _read_sample_metadata_comment(self, samples_dir: Path, file: str) -> Dict:
+        """
+        Additional meta-information can be provided through embedded comments:
+
+        // sample-metadata:
+        //   title: ACL (Access Control)
+        //   description: Demonstrates setting access control rules.
+        //   usage: node iam.js --help
+        """
+        sample_metadata = {}  # type: Dict[str, str]
+        with open(samples_dir / file) as f:
+            contents = f.read()
+            match = re.search(
+                r"(?P<metadata>// *sample-metadata:([^\n]+|\n//)+)", contents, re.DOTALL
+            )
+            if match:
+                # the metadata yaml is stored in a comments, remove the
+                # prefix so that we can parse the yaml contained.
+                sample_metadata_string = re.sub(
+                    r"((#|//) ?)", "", match.group("metadata")
+                )
+                sample_metadata = yaml.load(
+                    sample_metadata_string, Loader=yaml.SafeLoader
+                )["sample-metadata"]
+        return sample_metadata
+
+    def _read_quickstart(self, samples_dir: Path) -> str:
+        """
+        quickstart is a special case, it should be read from disk and displayed
+        in README.md rather than pushed into samples array.
+        """
+        reading = False
+        quickstart = ""
+
+        with open(samples_dir / "quickstart.js") as f:
+            while True:
+                line = f.readline()
+                if not line or re.search(_RE_SAMPLE_COMMENT_END, line):
+                    break
+                if reading:
+                    quickstart += line
+                if re.search(_RE_SAMPLE_COMMENT_START, line):
+                    reading = True
+
+        return quickstart
+
+    def _load_partials(self):
+        """
+        hand-crafted artisinal markdown can be provided in a .readme-partials.yml.
+        The following fields are currently supported:
+
+        body: custom body to include in the usage section of the document.
+        samples_body: an optional body to place below the table of contents
+          in samples/README.md.
+        introduction: a more thorough introduction than metadata["description"].
+        title: provide markdown to use as a custom title.
+        """
+        cwd_path = Path(os.getcwd())
+        partials_file = None
+        for file in [".readme-partials.yml", ".readme-partials.yaml"]:
+            if os.path.exists(cwd_path / file):
+                partials_file = cwd_path / file
+                break
+        if not partials_file:
+            return
+        with open(partials_file) as f:
+            self._data["partials"] = yaml.load(f, Loader=yaml.SafeLoader)
+
+    def __getitem__(self, key: str):
+        return self._data[key]
+
+
+class JavaMetadata(Metadata):
+    def __init__(self, initial_data: dict = {}):
+        Metadata.__init__(self, initial_data)
```

Comment:
Reusing code through inheritance is a known anti-pattern,
https://en.wikipedia.org/wiki/Composition_over_inheritance
further hinting that classes are a mismatch for this code.